anchore_engine/services/policy_engine/engine/vulns/stores.py

```diff
@@ -1,57 +1,108 @@
 import datetime
 import enum
-from dataclasses import dataclass
-from typing import Dict
+from dataclasses import dataclass, field
+from typing import Dict, Tuple
+
+import retrying
 
 from anchore_engine.apis.context import ApiRequestContextProxy
+from anchore_engine.clients.grype_wrapper import GrypeWrapperSingleton
 from anchore_engine.clients.services import internal_client_for, catalog
 from anchore_engine.db import (
     Image,
     get_thread_scoped_session as get_session,
-    CachedVulnerabilities,
+    ImageVulnerabilitiesReport as DbImageVulnerabilities,
 )
 from anchore_engine.services.policy_engine.api.models import ImageVulnerabilitiesReport
 from anchore_engine.subsys import logger as log
-from anchore_engine.clients.grype_wrapper import GrypeWrapperSingleton
 
-# Disabled by default, can be set in config file. Seconds for connection to cache
-DEFAULT_CACHE_CONN_TIMEOUT = -1
+# Disabled by default, can be set in config file. Seconds for connection to store
+DEFAULT_STORE_CONN_TIMEOUT = -1
 # Disabled by default, can be set in config file. Seconds for first byte timeout
-DEFAULT_CACHE_READ_TIMEOUT = -1
+DEFAULT_STORE_READ_TIMEOUT = -1
 
+REPORT_SAVE_RETRIES = 5  # number of save attempts
+REPORT_SAVE_WAIT = 1000  # wait between retries in milliseconds
 
-class CacheStatus(enum.Enum):
+
+class Status(enum.Enum):
     valid = "valid"
     stale = "stale"
     invalid = "invalid"
     missing = "missing"
 
 
-@dataclass
-class CacheRecord:
-    result: Dict
-    status: CacheStatus
+@dataclass(eq=True)
+class GrypeDBKey:
+    """
+    A key in the context of the image vulnerabilities store is the information that makes an entry in the store unique.
+    The key class has functions for generating an instance from the store's report_key and a ImageVulnerabilitiesReport
+    instance. It should also support a function for returning the status of the store entry
+
+    This key is based only on Grype DB details of the generated report.
 
+    To override the behaviour simply create a new key class and assign ImageVulnerabilitiesStore.__report_key_class__
+    to the new class
+    """
 
-@dataclass(eq=True)
-class GrypeMetadata:
-    version: str
-    db_version: str
+    # version: str
+    # db_version: str
     db_checksum: str
 
     @classmethod
-    def from_dict(cls, data):
-        return GrypeMetadata(
-            data.get("version"), data.get("db_version"), data.get("db_checksum")
+    def from_db(cls, report_key: Dict):
+        db_checksum = report_key.get("db_checksum") if report_key else None
+        if db_checksum:
+            return GrypeDBKey(db_checksum=db_checksum)
+        else:
+            raise ValueError(
+                "Invalid or unexpected report_key format {}".format(report_key)
+            )
+
+    @classmethod
+    def from_report(cls, report: ImageVulnerabilitiesReport):
```

Comment:
I think the db checksum that generated it should be included in the report itself as metadata and set when the report is generated. It can be used here as a key, but seems like its creation and lifecycle should live with the report itself.

Reply:
db checksum will be  accessed from the report but it's not currently available from grype. It will be with the next grype update. That's why the argument to the function is a report model object. This is a temporary workaround with a TODO right below :)

Reply:
Ah, ok it wasn't clear that ImageVulnerabilitiesReport should be raw and not include any wrappers or metadata envelope. I think it could with a metadata property and then the "matches" or similar. It would just be a slightly higher level abstraction, but if it's expected soon from Grype itself then can add it then.